### PR TITLE
[Networking] GossipSub Authorized Subscription

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR enables the peer scoring mechanism of GossipSub for Flow blockchain. The peer scoring mechanism at this phase only contains a subscription validator. When a peer discovers invalid subscriptions of other peers (based on their roles), it descends their local score to the `-Inf` value. With this approach, honest peers collectively refrain from routing the pubsub messages of (as well as to and from) malicious peers who conduct an invalid subscription. In other words, the malicious peers who conducted an invalid subscription are excluded from the GossipSub mesh of the honest peers.

### CODE REVIEW

File: network/p2p/scoring/subscription_validator_test.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	// mocks peer 1 not subscribed to an unknown topic.
+	peer1 := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer1).Return(unittest.IdentityFixture(), true)
+	sp.On("GetSubscribedTopics", peer1).Return([]string{"unknown-topic-1", "unknown-topic-2"})
+
+	// as peer 1 has subscribed to unknown topics, the subscription validator should return an error.
+	require.Error(t, sv.MustSubscribedToAllowedTopics(peer1))
```
Comment (peterargue) [inline network/p2p/scoring/subscription_validator_test.go:61]: if you make the unauthorized error a sentinel, you can check the error type here

File: network/p2p/scoring/subscription_validator_test.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		for i := range allowedTopics {
+			idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+			sp.On("GetSubscribedTopics", peer).Return(allowedTopics[:i+1])
+			require.NoError(t, sv.MustSubscribedToAllowedTopics(peer))
+		}
+	}
+}
+
+// TestSubscriptionValidator_SubscribeToAllTopics tests that regardless of its role when a peer has subscribed to all
+// topics, the subscription validator returns an error.
```
Comment (peterargue) [inline network/p2p/scoring/subscription_validator_test.go:94]: this is because within the list of all topics, there are guaranteed to be some that are not allowed for the given role? can you make that clear in the comment.

File: network/channels/channel_test.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// TestChannelList_ExcludeChannels tests that the channel list ExcludeChannels method returns a new channel list
+// with the excluded channels removed.
+func TestChannelList_ExcludeChannels(t *testing.T) {
+	list := channels.ChannelList{"a", "b", "c"}
+	require.ElementsMatch(t, channels.ChannelList{"a", "b"}, list.ExcludeChannels(channels.ChannelList{"c"}))
+	require.ElementsMatch(t, channels.ChannelList{"a", "c"}, list.ExcludeChannels(channels.ChannelList{"b"}))
+	require.ElementsMatch(t, channels.ChannelList{"b", "c"}, list.ExcludeChannels(channels.ChannelList{"a"}))
+	require.ElementsMatch(t, channels.ChannelList{"a", "b", "c"}, list.ExcludeChannels(channels.ChannelList{"d"}))
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/channels/channel_test.go:31]: https://github.com/onflow/flow-go/pull/3291/commits/1f83b205ece768eae006b741671dd5f8913f4a42

File: utils/slices/slices.go [hunk 4]
```diff
@@ -34,3 +34,29 @@ func MakeRange(min, max int) []int {
 	}
 	return a
 }
+
+// AreStringSlicesEqual returns true if the two string slices are equal.
+func AreStringSlicesEqual(a, b []string) bool {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply utils/slices/slices.go:41]: https://github.com/onflow/flow-go/pull/3291/commits/060e5163b4baeadeeb957d6144417cc58c36b228 

> It looks like this is only used for tests. Could you use assert.ElementsMatch() instead?

We need a test util that returns a boolean without failing the test [(example)](https://github.com/onflow/flow-go/blob/c3db15986b74dcfd7c9c7e7a1cb1dbcaa69dd977/network/p2p/scoring/subscription_provider_test.go#L46).

File: network/p2p/scoring/subscription_validator.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	topics := v.subscriptionProvider.GetSubscribedTopics(pid)
+
+	flowId, ok := v.idProvider.ByPeerID(pid)
+	if !ok {
+		return fmt.Errorf("could not find authorized identity for peer id %s", pid)
+	}
+
+	for _, topic := range topics {
+		if !p2putils.AllowedSubscription(flowId.Role, topic) {
+			return fmt.Errorf("unauthorized subscription: %s", topic)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator.go]: added in https://github.com/onflow/flow-go/pull/3291/commits/962f27b4fda15c4e41454a251355a31cd675be97 and https://github.com/onflow/flow-go/pull/3291/commits/f5d28dc7bc64e5730cb2f29a65ff0313f3f75f71

File: network/p2p/scoring/subscription_validator.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	return &SubscriptionValidator{
+		idProvider: idProvider,
+	}
+}
+
+func (v *SubscriptionValidator) RegisterSubscriptionProvider(provider p2p.SubscriptionProvider) {
+	v.subscriptionProvider = provider
+}
+
+// MustSubscribedToAllowedTopics validates all subscriptions a peer has with respect to all Flow topics.
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator.go:26]: https://github.com/onflow/flow-go/commit/f5d28dc7bc64e5730cb2f29a65ff0313f3f75f71

File: network/p2p/scoring/score_option.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		GossipThreshold:             DefaultGossipThreshold,
+		PublishThreshold:            DefaultPublishThreshold,
+		GraylistThreshold:           DefaultGraylistThreshold,
+		AcceptPXThreshold:           DefaultAcceptPXThreshold,
+		OpportunisticGraftThreshold: DefaultOpportunisticGraftThreshold,
+	}
+}
+
+func (s *ScoreOption) preparePeerScoreParams() {
+	s.peerScoreParams = &pubsub.PeerScoreParams{
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/score_option.go:121]: https://github.com/onflow/flow-go/pull/3291/commits/dd507ece77457c28f84976e2426182501dd38121

File: network/p2p/scoring/subscription_validator.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+func (v *SubscriptionValidator) MustSubscribedToAllowedTopics(pid peer.ID) error {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator.go]: https://github.com/onflow/flow-go/pull/3291/commits/0d3b21c66364e1c8d7162d7ab1605939bdf0b1c3

File: network/p2p/scoring/subscription_provider.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		allTopics: make([]string, 0),
+	}
+}
+
+// GetSubscribedTopics returns all the subscriptions of a peer within the pubsub network.
+// Note that the current peer must be subscribed to the topic for it to the same topics in order
+// to query for other peers, e.g., if current peer has subscribed to topics A and B, and peer1
+// has subscribed to topics A, B, and C, then GetSubscribedTopics(peer1) will return A and B. Since this peer
+// has not subscribed to topic C, it will not be able to query for other peers subscribed to topic C.
+func (s *SubscriptionProvider) GetSubscribedTopics(pid peer.ID) []string {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_provider.go:41]: This is called frequently upon every major event for a `peerId` in GossipSub, including but not limited to: accepting a connection, emitting gossips, sending protocol-level events like ihave and iwant, heartbeats (every 1 second by default), publishing a message, joining a topic, etc.

File: network/p2p/scoring/subscription_provider.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+func (s *SubscriptionProvider) getPeersByTopic(topic string) []peer.ID {
+	go func() {
+		// TODO: refactor this to a component manager worker once we have a startable libp2p node.
+		if _, updateInProgress := s.peersByTopicUpdating.LoadOrStore(topic, true); updateInProgress {
+			// another goroutine is already updating the list of peers for this topic
+			return
+		}
+
+		subscribedPeers := s.tp.ListPeers(topic)
+		s.peersByTopic.Store(topic, subscribedPeers)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_provider.go:103]: Currently, we (almost) only unsubscribe when we stop the node (at shutdown or restart). Hence, it is not deemed a frequent operation that causes leakage. Nevertheless, created this issue https://github.com/dapperlabs/flow-go/issues/6406 and added a todo in the code to refactor this later: https://github.com/onflow/flow-go/pull/3291/commits/d9d2b8811a1f6d76a8ee980e2c0256f1dca3fbd3

File: network/p2p/scoring/subscription_provider_test.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	}, 1*time.Second, 100*time.Millisecond)
+
+	assert.Eventually(t, func() bool {
+		return slices.AreStringSlicesEqual([]string{"topic1", "topic2"}, sp.GetSubscribedTopics(peer2))
+	}, 1*time.Second, 100*time.Millisecond)
+
+	assert.Eventually(t, func() bool {
+		return slices.AreStringSlicesEqual([]string{"topic2", "topic3"}, sp.GetSubscribedTopics(peer3))
+	}, 1*time.Second, 100*time.Millisecond)
+}
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_provider_test.go:48]: `tp` is initialized by the [default constructor](https://github.com/onflow/flow-go/blob/17ff050ec3110847ec60763759bdd9a23912c86c/network/p2p/scoring/subscription_provider_test.go#L20) of the testify mock, which has an `AssertExcpectations` call [embedded](https://github.com/onflow/flow-go/blob/7d5ee4cdc6170d00bcc351e11c0ac835a28fc82c/network/p2p/mock/topic_provider.go#L58).

File: network/p2p/scoring/subscription_validator_test.go (hunk 1 above)
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator_test.go]: https://github.com/onflow/flow-go/commit/f5d28dc7bc64e5730cb2f29a65ff0313f3f75f71

File: network/p2p/scoring/subscription_validator.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func (v *SubscriptionValidator) RegisterSubscriptionProvider(provider p2p.SubscriptionProvider) {
+	v.subscriptionProvider = provider
+}
+
+// CheckSubscribedToAllowedTopics validates all subscriptions a peer has with respect to all Flow topics.
+// All errors returned by this method are benign:
+// We expect an unauthorized or ejected peer to subscribe to Flow topics (InvalidPeerIDError).
+// We expect a peer to subscribe to a topic that is not allowed for its role (InvalidSubscriptionError).
```
Comment (peterargue) [inline network/p2p/scoring/subscription_validator.go:29]: I find these a little confusing. maybe rephrase to something like this?
```suggestion
// - InvalidPeerIDError: the subscribed peer is unauthorized (unknown) or ejected.
// - InvalidSubscriptionError: the peer is subscribed to a topic that is not allowed for its role.
```

File: network/p2p/scoring/subscription_validator_test.go [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+	allTopics := make([]string, 0, len(allChannels))
+	for _, channel := range allChannels {
+		allTopics = append(allTopics, channels.TopicFromChannel(channel, sporkID).String())
+	}
+
+	for _, role := range flow.Roles() {
+		peer := p2pfixtures.PeerIdFixture(t)
+		idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+		sp.On("GetSubscribedTopics", peer).Return(allTopics)
+		require.Error(t, sv.CheckSubscribedToAllowedTopics(peer), role)
```
Comment (peterargue) [inline network/p2p/scoring/subscription_validator_test.go:160]: can you check the specific expected error here

File: network/p2p/scoring/subscription_validator_test.go [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+		unauthorizedTopics := make([]string, 0, len(unauthorizedChannels))
+		for _, channel := range unauthorizedChannels {
+			unauthorizedTopics = append(unauthorizedTopics, channels.TopicFromChannel(channel, sporkID).String())
+		}
+
+		// peer should NOT pass subscription validator as it has subscribed to any subset of its unauthorized topics.
+		for i := range unauthorizedTopics {
+			idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+			sp.On("GetSubscribedTopics", peer).Return(unauthorizedTopics[:i+1])
+			require.Error(t, sv.CheckSubscribedToAllowedTopics(peer))
```
Comment (peterargue) [inline network/p2p/scoring/subscription_validator_test.go:188]: here as well

**Review by peterargue — APPROVED** [review]
added a few small comments, but otherwise looks good.

File: network/p2p/scoring/subscription_validator_test.go [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+
+	// as peer 1 has subscribed to unknown topics, the subscription validator should return an error.
+	err := sv.CheckSubscribedToAllowedTopics(peer1)
+	require.Error(t, err)
+	require.True(t, scoring.IsInvalidSubscriptionError(err))
+}
+
+// TestSubscriptionValidator_ValidSubscription tests that when a peer has subscribed to valid
+// topics based on its Flow protocol role, the subscription validator returns no error.
+func TestSubscriptionValidator_ValidSubscriptions(t *testing.T) {
```
Comment (kc1116) [inline network/p2p/scoring/subscription_validator_test.go:108]: Should we add one more test case where role is not a valid flow-role (observer nodes) can subscribe to public channels. Also ensure receive blocks can be subscribed to.

File: network/p2p/scoring/subscription_validator_test.go (hunk 13 above)
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator_test.go]: https://github.com/onflow/flow-go/pull/3291/commits/7c77d4b063f9316e53266c2ff886d4ad8fc756b6

File: network/p2p/scoring/subscription_validator_test.go (hunk 14 above)
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator_test.go]: https://github.com/onflow/flow-go/pull/3291/commits/2bab26414ed2bc3b91f74bb50d90e0272b5575d4

File: network/p2p/utils/p2putils.go [hunk 16]
```diff
@@ -69,3 +70,24 @@ func MultiAddressStr(ip, port string) string {
...
 }
+
+// AllowedSubscription returns true if the given role is allowed to subscribe to the topic.
+func AllowedSubscription(role flow.Role, topic string) bool {
+	channel, ok := channels.ChannelFromTopic(channels.Topic(topic))
+	if !ok {
+		return false
+	}
+
+	if !role.Valid() {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/utils/p2putils.go:81]: @vishalchangrani is observer node a valid staked (i.e., authorized) node that we expect on the Flow libp2p network?

[the remark this answers is not in the document]
Comment (vishalchangrani) [reply network/p2p/utils/p2putils.go:81]: So there are currently two Flow LibP2P network (check the image [here](https://developers.flow.com/nodes/node-operation/observer-node)). The observer node is not staked and should never be allowed to talk to anyone on the private network between collection, consensus, VN, AN and ENs.
However, please can you make sure that these changes don't break the ability of the observer to talk to the access node (the public Libp2p network).

[the remark this answers is not in the document]
Comment (vishalchangrani) [reply network/p2p/utils/p2putils.go:81]: observers currently are running with role `access`. Switching the observer to use the `observer` role is not that easy unfortunately and will need a lot of other changes (cc: @koko1123 and @peterargue )

File: network/p2p/scoring/subscription_validator_test.go (hunk 15 above)
Comment (yhassanzadeh13) [reply network/p2p/scoring/subscription_validator_test.go:108]: Good question, the peer scoring for GossipSub is currently activated only for the _staked_ (i.e., authorized) Flow network. The cardinal principle in the _staked_ Flow network is all nodes are having a valid role (VN, AN, EN, CON, or COL). We rely on the `IdentityProvider` in the sense that all resolved identities must be valid and authorized. By a resolved identity, we mean the one with a (`true`) returned value ([here](https://github.com/onflow/flow-go/blob/8e9601184f6385de5d1005cf111d47f406e0da82/network/p2p/scoring/subscription_validator.go#L33)). We consider it a [malicious action](https://github.com/onflow/flow-go/blob/8e9601184f6385de5d1005cf111d47f406e0da82/network/p2p/scoring/subscription_validator.go#L34-L40) if a libp2p `peer.ID` belonging to the _staked_ Flow network is not resolved to an authorized identity, and we penalize it with the maximum penalty value (so that it will be muted). 

Hence, the public Flow network remains unaffected by this PR.

File: network/p2p/utils/p2putils.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/utils/p2putils.go:81]: @vishalchangrani currently we enable GossipSub peer scoring only for the nodes built through [_scaffold_](https://github.com/onflow/flow-go/blob/ea4b8ae10634c1236dd96e2475f32d2b4a52367a/network/p2p/p2pbuilder/libp2pNodeBuilder.go#L83) (i.e., staked authorized roles).

At the current stage, we do not enable peer scoring for the following networks, as they would not affect the permissionless release of the _staked_ Flow network: 
- [Access Node Public Network](https://github.com/onflow/flow-go/blob/6f85281734f200cf0bb1fbf9662f62022d78f841/cmd/access/node_builder/access_node_builder.go#L977)
- [Follower Builder Service for Observer](https://github.com/onflow/flow-go/blob/4683340ec31171c2c1777886bd025be6bea0d368/follower/follower_builder.go#L567) (not sure about this, as it has the same documentation as the `ObserverServiceBuilder` cc @vishalchangrani @peterargue 
- [Observer Builder Service](https://github.com/onflow/flow-go/blob/6f85281734f200cf0bb1fbf9662f62022d78f841/cmd/observer/node_builder/observer_builder.go#L828). 

cc @kc1116 @vishalchangrani @peterargue

[the remark this answers is not in the document]
Comment (peterargue) [reply network/p2p/utils/p2putils.go:81]: Yes, those are the networks to avoid.

[the remark this answers is not in the document]
Comment (vishalchangrani) [reply network/p2p/utils/p2putils.go:81]: that sounds right ☝️

**Review by vishalchangrani — APPROVED** [review]

File: network/p2p/utils/p2putils.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/utils/p2putils.go:81]: @kc1116 on the _staked_ (i.e., authorized) Flow network at the current state, that code path is never invoked during GossipSub peer scoring process.
